presto-spi/src/main/java/com/facebook/presto/spi/block/MapBlockBuilder.java

```diff
@@ -504,6 +504,19 @@ private static int getHashPosition(Block keyBlock, int position, MethodHandle ke
             throw new RuntimeException(throwable);
         }
 
-        return (int) Math.floorMod(hashCode, hashTableSize);
+        return getReducedHashPosition(hashCode, hashTableSize);
+    }
+
+    // Adapted from the fastrange algorithm: https://github.com/lemire/fastrange. This function reduces the hashcode
+    // to [0, hashTableSize) uniformly. It first gets a 32 bit hashcode by calling Long.hashCode(), then reduce this 32
+    // bit hash code to [0, hashTableSize) by multiplying it by hashTableSize then divided by 2^32. By multiplying by
+    // hashTableSize, we take 32 bit hashcode in the range [0, 2^32) and map them to multiples of hashTableSize in
+    // [0, hashTableSize * 2^32). By dividing by 2^32, we map all multiples of hashTableSize in [0, 2^32) to 0, all
+    // multiples of hashTableSize in [2^32, 2 * 2^32) to one, and so forth. In each interval of 2^32 size, the number of
+    // reduced values must be 2^32/hashTableSize, therefore the distribution is uniform if the hashcode distribution is
+    // uniform.
+    static int getReducedHashPosition(long hashcode, int hashTableSize)
```

Comment:
nit: can this be private?

Reply:
This one is also used in SingleMapBlock.